griddedio/GriddedIO.F90

```diff
@@ -1013,6 +1013,10 @@ subroutine request_data_from_file(this,filename,timeindex,rc)
              this%read_collection_id, fileName, trim(names(i)), &
              & ref, start=localStart, global_start=globalStart, global_count=globalCount)
         deallocate(localStart,globalStart,globalCount)
+        ! if is Nan, do not set '_FillValue'.
+        ! The pair ESMF_AttributeSet and ESMF_AttributeGet cannot handle Nan
+        if (missing_value /= missing_value) cycle 
```

Comment:
Ok.   

1. Perhaps we should return an error if it is found to be NaN in this manner?   Better to fail here then when doing a regrid with such values later on.  
2. Please open a ticket with esmf support to see if this is something that they can fix on their end.
3. This mechanism for detecting NaN is just obscure enough that it is probably worth it to introduce a helper function `is_NaN(x)`  and then call that instead.   I would not ask, except you have to modify the code still to handle the error condition.

Reply:
If it is a "real" nan, `ieee_is_nan()` could do it for us. We use it in various places in GEOS. (Though for historical purposes we often rename it as `isnan`, a la https://cs.github.com/GEOS-ESM/GEOSgcm_GridComp/blob/3e52659a4baa55aea5b0c937afcbfe90f0fafd92/GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/Utils/mk_restarts/mk_CatchCNRestarts.F90#L199)

Reply:
The Nan now is changed to MAPL_UNDEF. But I think it is fine not to replace it.  The results will be Nan if a Nan is involved in calculation

Reply:
@weiyuan-jiang please elaborate on what you wrote above.    NaN is never MAPL_UNDEF, but I think you meant that the FillValue is now changed to MAPL_UNDEF.   This is not a change we are permitted to make though.

Reply:
I just don't know what the right thing to do here is, we could fail and return an error if the missing value is nan. On the other hand if we do not, the latest commit from weiyuan seems like all we can do. We can't let the nan's propagate to a real calculation. At least if the Nan's are replaced with MAPL_UNDEF which we using internal no matter what the missing value is those points will be skipped without leading to an error. I would approve but I will defer to @tclune.

Reply:
@tclune I am not sure about "This is not a change we are permitted to make though".  I think we simply replace Nan in the memory with MAPL_UNDEF for the convenience of calculation

Reply:
ok - trying to look at the code to see where this has an impact.  The CYCLE seems to suggest we don't need to tell ESMF the fill value in the first place.

Reply:
How about saving the metadata for swap call when request_data_from_file. Is it safe?  @bena-nasa  , please see my new change